### PR TITLE
Made circulation pump a branch component

### LINKED ISSUE #416: Changes to circulation pump component
**Problem description*
The circulation pumps currently include a few workarounds that should be solved.

**Solution**
This includes the following steps:
- [x] create-function should check integrity of type and given values -> if "p" is chosen, p_bar must be given, if "t" is chosen, t_k must be given -> the parameter type="auto" (check from p/t values) would be nice to have
- [x] The component should be made a branch component, so that the toolbox functions can identify circulation pumps correctly. There is currently some bad behavior or workarounds due to the fact that CircPump inherits from ExtGrid (which is a node element).

### PR DESCRIPTION
**Pull request by dlohmeier**
fixes #416

### CODE REVIEW

**Comment by dlohmeier** [comment]
We should still implement the rest of #416

**Comment by dlohmeier** [comment]
fixes #416

**Review by SimonRubenDrauz — APPROVED** [review]

File: pandapipes/component_models/abstract_models/circulation_pump.py [hunk 1]
```diff
@@ -16,14 +20,80 @@
...
+
+    @classmethod
+    def get_result_table(cls, net):
+        """
+
+        :param net: The pandapipes network
+        :type net: pandapipesNet
+        :return: (columns, all_float) - the column names and whether they are all float type. Only
+                if False, returns columns as tuples also specifying the dtypes
+        :rtype: (list, bool)
```
Comment (SimonRubenDrauz) [inline pandapipes/component_models/abstract_models/circulation_pump.py:41]: Looks good to me

File: pandapipes/test/api/test_components/test_ext_grid.py [hunk 2]
```diff
@@ -22,11 +22,9 @@ def test_ext_grid_sorting(use_numba):
...
     j6 = pandapipes.create_junction(net, 1, 293.15, index=7)
 
-    pandapipes.create_ext_grid(net, j2, 1, 285.15, type="pt")
-    pandapipes.create_ext_grid(net, j3, 1, 285.15, type="pt")
-    pandapipes.create_ext_grid(net, j5, 1, 285.15, type="t")
-    pandapipes.create_ext_grid(net, j1, 1, 285.15, type="pt")
-    pandapipes.create_ext_grid(net, j1, 1, 285.15, type="pt")
+    pandapipes.create_ext_grids(net, [j2, j3, j5, j1, j1], p_bar=1, t_k=285.15,
+                                type=["auto", "pt", "t", "pt", "tp"])
+    assert np.all(net.ext_grid.type == ["pt", "pt", "t", "pt", "pt"])
```
Comment (SimonRubenDrauz) [inline pandapipes/test/api/test_components/test_ext_grid.py:27]: Maybe you should also add a test, where some ps and ts are None. 

Furthermore, what seems a bit strange to me. If I define p and have a t ext grid, the p wouldn't have any effect. 

Therefore, I would set p and t by default equals to None (see my ext_grid comment). Also there should be some kind of warning. If you set p and say the type is 't', a warning should pop up saying something like p has no effect.

File: pandapipes/plotting/simple_plot.py [hunk 3]
```diff
@@ -284,24 +284,24 @@ def create_simple_collections(net, respect_valves=False, respect_in_service=True
             circ_pump_colls = create_pump_collection(
                 net, pumps=net.circ_pump_mass[net.circ_pump_mass.in_service].index,
                 table_name='circ_pump_mass', size=pump_size, linewidths=pipe_width,
-                color=pump_color)
+                color=pump_color, fj_col="return_junction", tj_col="flow_junction")
```
Comment (SimonRubenDrauz) [inline pandapipes/plotting/simple_plot.py:287]: Would it make sense instead of passing the junction names, to pass the components itself, as the components have the corresponding class method. So in case we change the name again, there is no need of adaptiong the code here.

File: pandapipes/plotting/generic_geodata.py [hunk 4]
```diff
@@ -53,19 +51,18 @@ def build_igraph_from_ppipes(net, junctions=None):
...
-        g.add_edge(pp_junction_mapping[junction.from_junction],
-                   pp_junction_mapping[junction.to_junction],
-                   weight=junction.length_km)
+    for pipe in net.pipe[mask].itertuples():
+        g.add_edge(pp_junction_mapping[pipe.from_junction], pp_junction_mapping[pipe.to_junction],
+                   weight=pipe.length_km)
 
     for comp in net['component_list']:
-        if comp in [Source, Sink, ExtGrid, Pipe, Junction]:
+        if not isinstance(comp, (BranchComponent, CirculationPump)):
```
Comment (SimonRubenDrauz) [inline pandapipes/plotting/generic_geodata.py:59]: Why is this necessary? I thought, CirculationPumps are BranchComponents as well now.

File: pandapipes/plotting/generic_geodata.py [hunk 5]
```diff
@@ -53,19 +51,18 @@ def build_igraph_from_ppipes(net, junctions=None):
...
             continue
         mask = _get_element_mask_from_nodes(
-            net, comp.table_name(), ["from_junction", "to_junction"], junctions)
+            net, comp.table_name(), comp.from_to_node_cols(), junctions)
+        fjc, tjc = comp.from_to_node_cols()
```
Comment (SimonRubenDrauz) [inline pandapipes/plotting/generic_geodata.py:63]: Put this first, then you do not need to call the function twice.

File: pandapipes/io/convert_format.py [hunk 6]
```diff
@@ -42,6 +42,13 @@ def _rename_columns(net):
             else:
                 net['controller'].drop('controller', inplace=True, axis=1)
         net["controller"].rename(columns={"controller": "object"}, inplace=True)
+    for cp_tbl in ["circ_pump_mass", "circ_pump_pressure"]:
+        if cp_tbl in net:
+            for old_col, new_col in [("from_junction", "flow_junction"),
+                                     ("to_junction", "return_junction")]:
+                if old_col in net[cp_tbl] and new_col not in net[cp_tbl]:
+                    net[cp_tbl][new_col] = net[cp_tbl][old_col]
+                    net[cp_tbl].drop(old_col, inplace=True, axis=1)
```
Comment (SimonRubenDrauz) [inline pandapipes/io/convert_format.py:51]: why not using the rename functionality of pandas? Maybe, it would make sense to pass the component here as well. By that, if the name changes again, you do not need to adapt this as well.

File: pandapipes/component_models/circulation_pump_mass_component.py [hunk 7]
```diff
@@ -46,24 +66,12 @@ def create_pit_node_entries(cls, net, node_pit):
         mf = np.nan_to_num(circ_pump.mdot_kg_per_s.values)
         mass_flow_loads = mf * circ_pump.in_service.values
         juncts, loads_sum = _sum_by_group(get_net_option(net, "use_numba"),
-                                          circ_pump.to_junction.values, mass_flow_loads)
+                                          circ_pump.return_junction.values, mass_flow_loads)
         junction_idx_lookups = get_lookup(net, "node", "index")[
             cls.get_connected_node_type().table_name()]
         index = junction_idx_lookups[juncts]
         node_pit[index, LOAD] += loads_sum
```
Comment (SimonRubenDrauz) [inline pandapipes/component_models/circulation_pump_mass_component.py:73]: If I compare this to the other circulation pump I have the feeling, that this needs to be adapted. Following things as example should be revised:

1. Why is it not mdot_flow_kg_per_s
2. As this is branch_component now, the mass flow should not be part of the node_pit, as in fact there is no load at the return_junction.

**Review by nsanina — APPROVED** [review]

File: pandapipes/component_models/abstract_models/branch_wzerolength_models.py [hunk 8]
```diff
@@ -54,11 +54,11 @@ def create_pit_branch_entries(cls, net, branch_pit):
...
         """
-        branch_wizerolength_pit = \
+        branch_wzerolength_pit = \
             super().create_pit_branch_entries(net, branch_pit)
-        branch_wizerolength_pit[:, LENGTH] = 0
-        branch_wizerolength_pit[:, K] = 1000
-        return branch_wizerolength_pit
+        branch_wzerolength_pit[:, LENGTH] = 0
+        branch_wzerolength_pit[:, K] = 1000
+        return branch_wzerolength_pit
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/component_models/abstract_models/branch_wzerolength_models.py:61]: just to make it fit the class name (better readability)

File: pandapipes/component_models/abstract_models/branch_models.py [hunk 9]
```diff
@@ -87,8 +87,9 @@ def create_pit_branch_entries(cls, net, branch_pit):
 
         junction_idx_lookup = get_lookup(net, "node", "index")[
             cls.get_connected_node_type().table_name()]
-        from_nodes = junction_idx_lookup[net[cls.table_name()]["from_junction"].values]
-        to_nodes = junction_idx_lookup[net[cls.table_name()]["to_junction"].values]
+        fn_col, tn_col = cls.from_to_node_cols()
+        from_nodes = junction_idx_lookup[net[cls.table_name()][fn_col].values]
+        to_nodes = junction_idx_lookup[net[cls.table_name()][tn_col].values]
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/component_models/abstract_models/branch_models.py:92]: Because the `cls.from_to_node_cols()` is more generic and should always apply, even if the columns are renamed again.

File: pandapipes/component_models/abstract_models/circulation_pump.py [hunk 10]
```diff
@@ -43,30 +113,34 @@ def extract_results(cls, net, options, branch_results, nodes_connected, branches
...
-        res_table, circ_pump, index_nodes_from, node_pit, _ = \
-            super().extract_results(net, options, None, nodes_connected, branches_connected)
+        circ_pump_tbl = net[cls.table_name()]
 
-        index_juncts_to = circ_pump.to_junction.values
-        junct_uni_to = np.array(list(set(index_juncts_to)))
-        index_nodes_to = get_lookup(net, "node", "index")[
-            cls.get_connected_node_type().table_name()][junct_uni_to]
+        if len(circ_pump_tbl) == 0:
+            return
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/component_models/abstract_models/circulation_pump.py:119]: There have been a lot of changes to this function, which is why it's not so easy to follow up.
1. Before, circulation pump inherited from ext_grid, and the largest part of the result extraction was done there. Now, it is all implemented in the circulation pump itself.
2. One functionality has been externalized (`get_mass_flow_at_nodes`), as it is used by different components. It performs a mass flow balance around fixed p nodes to identify the missing flow.
Line 118/119 just state that in case of an empty table, no results shall be extracted to avoid errors to be raised just due to facts like elements being dropped in the pandas table, without deleting the component from the component list.

File: pandapipes/component_models/abstract_models/circulation_pump.py [hunk 11]
```diff
@@ -43,30 +113,34 @@ def extract_results(cls, net, options, branch_results, nodes_connected, branches
...
 
-        deltap_bar = node_pit[index_nodes_from, PINIT] - node_pit[index_nodes_to, PINIT]
-        res_table["deltap_bar"].values[:] = deltap_bar
+        res_table = net["res_" + cls.table_name()]
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/component_models/abstract_models/circulation_pump.py:121]: No, this is just related to how the diff is shown here. The function has changed completely.

File: pandapipes/component_models/abstract_models/circulation_pump.py [hunk 12]
```diff
@@ -43,30 +113,34 @@ def extract_results(cls, net, options, branch_results, nodes_connected, branches
...
-        return ["mdot_kg_per_s", "deltap_bar"], True
+        # positive results mean that the circ_pump feeds in, negative means that the ext grid
+        # extracts (like a load)
+        res_table["mdot_flow_kg_per_s"].values[p_grids] = - (sum_mass_flows / counts)[inverse_nodes]
+
+        return_junctions = circ_pump_tbl[fn_col].values
+        return_nodes = junction_lookup[return_junctions]
+
+        deltap_bar = node_pit[flow_nodes, PINIT] - node_pit[return_nodes, PINIT]
+        res_table["deltap_bar"].values[in_service] = deltap_bar[in_service]
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/component_models/abstract_models/circulation_pump.py:146]: This is again due to how the diff is displayed. I also reorganized the structure of the circulation_pump class, so that the methods appear in more or less the same order in all components.

File: pandapipes/component_models/circulation_pump_mass_component.py (hunk 7 above)
Comment (dlohmeier) [reply pandapipes/component_models/circulation_pump_mass_component.py:73]: For the first comment, I understand that you would prefer to change the column name? That would be fine to me, just requires another API change, like for flow / return junction.

With respect to the node_pit, I believe that this is the only way to implement the circulation pump in our current Jacobi-matrix setup. I believe that we will definitely end up with a new structure in future, where each component can feed its own model right into the jacobi matrix. But this would be a long way to go still, and maybe it should be a step to take afterwards.

File: pandapipes/io/convert_format.py [hunk 13]
```diff
@@ -42,6 +42,13 @@ def _rename_columns(net):
             else:
                 net['controller'].drop('controller', inplace=True, axis=1)
         net["controller"].rename(columns={"controller": "object"}, inplace=True)
+    for cp_tbl in ["circ_pump_mass", "circ_pump_pressure"]:
+        if cp_tbl in net:
+            for old_col, new_col in [("from_junction", "flow_junction"),
+                                     ("to_junction", "return_junction")]:
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/io/convert_format.py:48]: Did we do that with other components, or was there a need to do something like that before? I don't really see a reason why we should not do that, and it might make life a little easier, that's true.

File: pandapipes/io/convert_format.py (hunk 6 above)
Comment (dlohmeier) [reply pandapipes/io/convert_format.py]: Did that in the new commit.

File: pandapipes/io/convert_format.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/io/convert_format.py]: Added this in the new commit.

File: pandapipes/plotting/generic_geodata.py (hunk 4 above)
Comment (dlohmeier) [reply pandapipes/plotting/generic_geodata.py]: Removed that.

File: pandapipes/plotting/generic_geodata.py (hunk 5 above)
Comment (dlohmeier) [reply pandapipes/plotting/generic_geodata.py:61]: Did that.

File: pandapipes/create.py [hunk 14]
```diff
@@ -204,8 +205,8 @@ def create_source(net, junction, mdot_kg_per_s, scaling=1., name=None, index=Non
     return index
 
 
-def create_ext_grid(net, junction, p_bar, t_k, name=None, in_service=True, index=None, type="pt",
-                    **kwargs):
+def create_ext_grid(net, junction, p_bar, t_k, name=None, in_service=True, index=None, type="auto",
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/create.py:208]: You mean the auto type detection? That would be a little tricky, as I also need it for the circulation pump, and then it requires the other PR to be accepted and merged before I can do that.

File: pandapipes/create.py [hunk 15]
```diff
@@ -631,18 +640,24 @@ def create_pump_from_parameters(net, from_junction, to_junction, new_std_type_na
     return index
 
 
-def create_circ_pump_const_pressure(net, from_junction, to_junction, p_bar, plift_bar,
-                                    t_k=None, name=None, index=None, in_service=True, type="pt",
-                                    **kwargs):
+def create_circ_pump_const_pressure(net, return_junction, flow_junction, p_bar, plift_bar,
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/create.py:643]: I can understand the intuition behind this. Yet, when considering the normal branch model, you would always define the ingoing junction first, then the outgoing one. That is why I implemented it this way..

File: pandapipes/test/api/test_components/test_ext_grid.py (hunk 2 above)
Comment (dlohmeier) [reply pandapipes/test/api/test_components/test_ext_grid.py:27]: I did add the check for non-null p or t values.

Comment (dlohmeier) [reply pandapipes/test/api/test_components/test_ext_grid.py:27]: About the None default, I am not so sure. I thought about that and didn't come to a clear conclusion, as both is not exactly user-friendly: on the one hand, giving many default values and expecting one to be set, on the other hand giving some positional arguments and expecting the user to know that in some cases one of them should be set to None (which would be an argument for a default value).

File: pandapipes/create.py [hunk 16]
```diff
@@ -204,8 +205,8 @@ def create_source(net, junction, mdot_kg_per_s, scaling=1., name=None, index=Non
     return index
 
 
-def create_ext_grid(net, junction, p_bar, t_k, name=None, in_service=True, index=None, type="pt",
-                    **kwargs):
+def create_ext_grid(net, junction, p_bar, t_k, name=None, in_service=True, index=None, type="auto",
+                    check_types=True, **kwargs):
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/create.py]: Removed the argument, type checks are always performed. The "auto" type will always be overridden.

File: pandapipes/create.py [hunk 17]
```diff
@@ -251,6 +257,9 @@ def create_ext_grid(net, junction, p_bar, t_k, name=None, in_service=True, index
     _check_junction_element(net, junction)
     index = _get_index_with_check(net, "ext_grid", index, name="external grid")
 
+    if check_types:
+        type = _auto_ext_grid_type(p_bar, t_k, type, ExtGrid)
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/create.py]: see above

File: pandapipes/create.py [hunk 18]
```diff
@@ -1321,3 +1423,143 @@ def _check_std_type(net, std_type, table, function_name):
     if std_type not in net['std_types'][table]:
         raise UserWarning('%s is not given in std_types (%s). Either change std_type or define new '
                           'one' % (std_type, table))
+
+
+ALLOWED_EG_TYPES = ["auto", "t", "p", "pt", "tp"]
+
+
+def _auto_ext_grid_type(p_bar, t_k, typ, comp):
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/create.py:1414]: I think that the toolbox is a set of functions for the user to just make use of to change stuff in the net (actually it is a little overloaded, but I think that it should be the intention behind it). This function is quite dedicated to the create-module, so I would prefer to leave it there (like some internal functions in the pandapower create module).

File: pandapipes/create.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/create.py:208]: see my other comment, I can't exactly tell which version (with or without default values) is really better. Maybe we should discuss that again.

File: pandapipes/create.py [hunk 19]
```diff
@@ -4,6 +4,7 @@
 
 import numpy as np
 import pandas as pd
+from numpy import ma
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/create.py]: Done

File: pandapipes/create.py [hunk 20]
```diff
@@ -1321,3 +1423,143 @@ def _check_std_type(net, std_type, table, function_name):
...
+        p_bar = np.array([p_bar] * length)
+
+    if not t_arr:
+        t_k = np.array([t_k] * length)
+
+    if not typ_arr:
+        typ = np.array([typ] * length)
+
+    p_null = np.equal(p_bar, None) | np.isnan(p_bar)
+    t_null = np.equal(t_k, None) | np.isnan(t_k)
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/create.py:1515]: This is one of the bigger issues with the bulk create functions. For a user, there are usually some preceding steps prior to the function call. I am not exactly sure if we will always end up with an array or a list containing None's in all cases. It's hard to anticipate what to expect there.

File: pandapipes/create.py [hunk 21]
```diff
@@ -1321,3 +1423,143 @@ def _check_std_type(net, std_type, table, function_name):
...
+        invalid_ind = np.where(ptn)
+        raise UserWarning("For component %s in positions %s, either pressure or temperature must be"
+                          " defined!" % (comp.__name__, invalid_ind))
+
+    if np.any(~np.isin(typ, ALLOWED_EG_TYPES)):
+        raise UserWarning("The types for component %s are %s, but must be one of the following: %s."
+                          % (comp.__name__, typ, ALLOWED_EG_TYPES))
+
+    p_types = np.isin(typ, [tp for tp in ALLOWED_EG_TYPES if tp != "t"])
+    t_types = np.isin(typ, [tp for tp in ALLOWED_EG_TYPES if tp != "p"])
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/create.py]: I just did not consider auto here, would prefer to do the checks first.

File: pandapipes/io/convert_format.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapipes/io/convert_format.py]: I don't think, that we did this before. It just came to my mind and as you also stated, it might make life easier. We should keep this in mind though, in case we stumble upon a similiar case.

File: pandapipes/component_models/circulation_pump_mass_component.py (hunk 7 above)
Comment (SimonRubenDrauz) [reply pandapipes/component_models/circulation_pump_mass_component.py:73]: 1. I would do it due to consistency reasons.
2. Agree. First things first.

File: pandapipes/create.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapipes/create.py:1515]: But what happens in the non-bulk function. If I as a user set p_bar=np.nan is this possible?

File: pandapipes/create.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapipes/create.py:207]: Agree.

File: pandapipes/create.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapipes/create.py:1413]: I see. In another PR about the mixture calculation I introduced a create_toolbox. Maybe we should put a TODO as soon as we merge the mixture calculation and put these functions there as I have the feeling that create.py as well as toolbox.py should only contain functions a user needs to interact with the net. What do you think?

File: pandapipes/create.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapipes/create.py:207]: No, actually I thought to put the changes of ext_grid in a seperate PR. But is find for me now.

File: pandapipes/create.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapipes/create.py]: Okay, I see. Then I would say, leave it like that. As soon as a discussion pops up, we could rediscuss this point.

File: pandapipes/create.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapipes/create.py]: I don't understand what you mean by 'see above'. What are you refering to?

File: pandapipes/create.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapipes/create.py]: Great, thanks!

File: pandapipes/create.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (SimonRubenDrauz) [reply pandapipes/create.py]: Great!

File: pandapipes/test/api/test_components/test_ext_grid.py (hunk 2 above)
Comment (SimonRubenDrauz) [reply pandapipes/test/api/test_components/test_ext_grid.py:27]: Really difficult question. Maybe we can discuss this again.

File: tutorials/circular_flow_in_a_district_heating_grid.ipynb [hunk 22]
```diff
@@ -71,7 +71,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "pp.create_circ_pump_const_mass_flow(net, from_junction=j0, to_junction=j3, p_bar=5,\n",
+    "pp.create_circ_pump_const_mass_flow(net, return_junction=j0, flow_junction=j3, p_bar=5,\n",
     "                                    mdot_kg_per_s=20, t_k=273.15+35)"
    ]
   },
```
[the remark this answers is not in the document]
Comment (dlohmeier) [reply tutorials/circular_flow_in_a_district_heating_grid.ipynb]: Thanks for the hint, I did that.

File: pandapipes/create.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/create.py:1413]: Sounds good to me. Maybe we can suggest that for pandapower as well.

File: pandapipes/create.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (dlohmeier) [reply pandapipes/create.py:1522]: I changed it so that the behavior in the bulk and non-bulk functions are the same and both accept NaN as None.

File: pandapipes/test/api/test_components/test_ext_grid.py (hunk 2 above)
Comment (dlohmeier) [reply pandapipes/test/api/test_components/test_ext_grid.py:27]: Solution: For ext_grids, None is used as default for p_bar and t_k, the type-keyword was moved a little further up front. For circulation pumps, the p_bar value must still be given and t_k remains None by default, also here the type keyword was moved a little further up front.

**Comment by SimonRubenDrauz** [comment]
Opened a new PR due to merge conflicts

**Comment by SimonRubenDrauz** [comment]
#449
